cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -332,49 +301,250 @@ private void addWorkflowNodeState(ProgramRunId programRunId, Map<String, String>
+    MDSKey.Builder key = getProgramKeyBuilder(TYPE_RUN_RECORD_COMPLETED, programRunId.getParent());
+
+    RunRecordMeta existing = getRun(programRunId);
+    if (existing == null) {
+      LOG.debug("Ignoring unexpected transition of program run {} to cluster state {} with no existing run record.",
+                programRunId, ProgramRunClusterStatus.PROVISIONED);
+      return null;
+    } else if (!isValid(existing, sourceId, "provisioned")) {
+      return null;
+    }
+
+    ProgramRunClusterStatus clusterStatus = existing.getCluster().getStatus();
+    if (clusterStatus != ProgramRunClusterStatus.PROVISIONED) {
+      LOG.warn("Ignoring unexpected request to transition program run {} from cluster state {} to cluster state {}.",
+               programRunId, clusterStatus, ProgramRunClusterStatus.PROVISIONED);
+      return null;
+    }
+    if (!existing.getStatus().isEndState()) {
+      LOG.warn("Ignoring unexpected request to transition program run {} from program state {} to cluster state {}.",
+               programRunId, existing.getStatus(), ProgramRunClusterStatus.PROVISIONED);
+      return null;
+    }
+
+    key.add(getInvertedTsKeyPart(existing.getStartTs())).add(programRunId.getRun()).build();
+
+    ProgramRunCluster cluster = new ProgramRunCluster(ProgramRunClusterStatus.DEPROVISIONING,
+                                                      null, existing.getCluster().getNumNodes());
+    RunRecordMeta meta = RunRecordMeta.builder(existing)
+      .setCluster(cluster)
+      .setSourceId(sourceId)
+      .build();
+    write(key.build(), meta);
+    return ProgramRunClusterStatus.DEPROVISIONING;
+  }
+
+  /**
+   * Record that the program run has deprovisioned compute resources for the run. If the current status has
+   * a higher source id, this call will be ignored.
+   *
+   * @param programRunId program run
+   * @param sourceId unique id representing the source of program run status, such as the message id of the program
+   *                 run status notification in TMS. The source id must increase as the recording time of the program
+   *                 run status increases, so that the attempt to persist program run status older than the existing
+   *                 program run status will be ignored
+   * @return {@link ProgramRunClusterStatus#DEPROVISIONED} if it is successfully persisted, {@code null} otherwise.
+   */
+  @Nullable
+  public ProgramRunClusterStatus recordProgramDeprovisioned(ProgramRunId programRunId, byte[] sourceId) {
+    MDSKey.Builder key = getProgramKeyBuilder(TYPE_RUN_RECORD_COMPLETED, programRunId.getParent());
+
+    RunRecordMeta existing = getRun(programRunId);
+    if (existing == null) {
+      LOG.debug("Ignoring unexpected transition of program run {} to cluster state {} with no existing run record.",
+                programRunId, ProgramRunClusterStatus.PROVISIONED);
```

Comment:
do you mean `ProgramRunClusterStatus.DEPROVISIONED`?

Reply:
oops, you're right